File: netlify/functions/vectorizar.py
import json
import urllib.request

# Modelo gratuito de Hugging Face (no necesitas pagar nada)
API_URL = "https://api-inference.huggingface.co/models/sentence-transformers/all-MiniLM-L6-v2"

# NOTA: Hugging Face ofrece una capa gratuita pública. 
# Si haces demasiadas peticiones seguidas, te conviene crear un Token gratis en su web 
# y ponerlo en las variables de entorno de Netlify como HF_TOKEN.
import os
HF_TOKEN = os.environ.get("HF_TOKEN", "") 

def handler(event, context):
    if event.get("httpMethod") != "POST":
        return {
            "statusCode": 405,
            "body": json.dumps({"error": "Método no permitido. Usa POST."})
        }
    
    try:
        body = json.loads(event.get("body", "{}"))
        texto = body.get("texto")
        
        if not texto:
            return {
                "statusCode": 400,
                "body": json.dumps({"error": "Falta el campo 'texto'."})
            }
        
        # Preparar la petición a Hugging Face
        payload = json.dumps({"inputs": texto}).encode("utf-8")
        headers = {"Content-Type": "application/json"}
        if HF_TOKEN:
            headers["Authorization"] = f"Bearer {HF_TOKEN}"
            
        req = urllib.request.Request(API_URL, data=payload, headers=headers)
        
        # Enviar y recibir la respuesta
        with urllib.request.urlopen(req) as response:
            res_body = response.read().decode("utf-8")
            embedding = json.loads(res_body)
            
        return {
            "statusCode": 200,
            "headers": {
                "Content-Type": "application/json",
                "Access-Control-Allow-Origin": "*"
            },
            "body": json.dumps({
                "texto_original": texto,
                "dimensiones": len(embedding) if isinstance(embedding, list) else None,
                "vector": embedding
            })
        }
        
    except Exception as e:
        return {
            "statusCode": 500,
            "body": json.dumps({"error": str(e)})
        }

File: netlify/functions/vectorizar.js
exports.handler = async (event, context) => {
  if (event.httpMethod !== "POST") {
    return {
      statusCode: 405,
      body: JSON.stringify({ error: "Método no permitido. Usa POST." }),
    };
  }

  try {
    const body = JSON.parse(event.body || "{}");
    const texto = body.texto;

    if (!texto) {
      return {
        statusCode: 400,
        body: JSON.stringify({ error: "Falta el campo 'texto' en el JSON." }),
      };
    }

    // Importación dinámica de la librería de Transformers de Hugging Face
    const { pipeline } = await import('@xenova/transformers');

    // Inicializamos el extractor de características con el modelo ligero
    const extractor = await pipeline('feature-extraction', 'Xenova/all-MiniLM-L6-v2');

    // Generamos la vectorización (embedding)
    const output = await extractor(texto, { pooling: 'mean', normalize: true });
    
    // Convertimos el resultado a un array común de JavaScript
    const embedding = Array.from(output.data);

    return {
      statusCode: 200,
      headers: {
        "Content-Type": "application/json",
        "Access-Control-Allow-Origin": "*",
      },
      body: JSON.stringify({
        texto_original: texto,
        dimensiones: embedding.length,
        vector: embedding,
      }),
    };

  } catch (error) {
    return {
      statusCode: 500,
      body: JSON.stringify({ error: error.message }),
    };
  }
};
